### ISSUE TITLE
[RFC]: KV-Gather CP and persistent DiT gap parallelism

### ISSUE THREAD

**Issue by nussejzz** [vllm-project/vllm-omni#5314]
## Motivation

We propose a reusable parallelism feature for Diffusion Transformer (DiT)
models, consisting of two composable parts:

1. **KV-Gather Context Parallelism (CP)** keeps Q and attention output local,
   All-Gathers K/V once per layer, and overlaps communication with independent
   K/V/Q projections.
2. **DiT gap parallelism** preserves the sequence shard outside transformer
   blocks and across denoising steps, so the DiT, CFG, scheduler update, and the
   next DiT invocation remain sharded.

Together they provide persistent sharding over the complete DiT denoising
process. The feature is intended to be available to all compatible DiT models;
BAGEL is the first integration and validation example, not the scope of the
design.

DiT inference is currently sequence-parallelized primarily with Ulysses or Ring
Attention. Ulysses communicates Q, K, and V before attention and communicates
the output afterwards. Ring Attention keeps Q local, but circulates K/V blocks
through the ranks and incrementally combines them with online softmax.

KV-Gather is particularly useful for grouped-query attention (GQA) and
multi-query attention (MQA), where K/V are much smaller than Q and the output.
BAGEL demonstrates this with 28 query heads and 4 key/value heads.

### 1. KV-Gather attention

Split the sequence over `P` ranks. Rank `i` owns `(Q_i, K_i, V_i)`.

Ring Attention keeps `Q_i` local and receives each `(K_j, V_j)` block in turn.
The only cross-block reduction required by attention is softmax. Ring Attention
therefore maintains local online-softmax state (row maximum, exponential sum,
and output accumulator) while K/V blocks circulate. The online-softmax state
normally remains on the Q-owning rank; the ring transfers K/V.

KV-Gather CP starts from the same decomposition but replaces iterative K/V
circulation with one collective:

```text
K = AllGather(K_i) = [K_0, ..., K_(P-1)]
V = AllGather(V_i) = [V_0, ..., V_(P-1)]
O_i = Attention(Q_i, K, V)
```

Both approaches expose every local query to the same global K/V sequence and
are mathematically equivalent to the corresponding slice of unsharded
attention. KV-Gather materializes full K/V and can invoke a regular attention
backend, so it does not need iterative ring transfers or cross-block
online-softmax state. Different softmax reduction orders and attention backends
can still prevent bitwise equivalence.

KV-Gather does not fundamentally require equal local sequence lengths. A
general implementation caches the valid length of each rank, pads local K/V to
the maximum shard length for the asynchronous collective, and removes padding
before concatenating global K/V. Local Q remains unpadded and can have a
different length on every rank. Since padding is removed before attention, this
does not require an attention mask. Shard-length metadata is stable for a DiT
shape and should be exchanged once rather than once per layer.

### 2. Communication volume

Let:

- `P` be the sequence-parallel degree;
- `H_q` be the number of query heads;
- `H_kv` be the number of key/value heads;
- `R = H_q / H_kv` be the GQA head ratio.

Assume equal sequence shards, `H_k = H_v`, equal Q/output head width, and ignore
collective-algorithm overhead. Counting per-rank remote payload:

```text
KV-Gather:
V_CP = 2 * (P - 1) * (S / P) * H_kv * D

Ulysses:
V_UP = ((P - 1) / P) * (S / P) *
       (H_q + H_kv + H_kv + H_q) * D

V_CP / V_UP = P * H_kv / (H_q + H_kv)
            = P / (R + 1)
```

For BAGEL, `H_q=28`, `H_kv=4`, and `R=7`. At degree 4:

```text
V_CP / V_UP = 4 / (7 + 1) = 1/2
```

KV-Gather communicates less when `P < R+1`, the same amount when `P = R+1`,
and more when `P > R+1`. This makes the design most attractive for GQA/MQA at
moderate parallel degrees. The formula describes balanced latent shards;
variable-length shards have rank-specific payloads plus any temporary padding
overhead, while preserving the same head-ratio comparison.

The tradeoff is memory: every rank stores the complete K/V tensors for the
layer. KV-Gather exchanges memory capacity for lower communication volume and a
simple collective; it does not eliminate the K/V communication itself.

### 3. Asynchronous projection/communication overlap

The reusable CP path allows a model's Q/K/V projection to expose independent
components, even when their weights are stored in one packed projection. It
schedules them in this order:

```text
K projection
  -> async K All-Gather
V projection
  -> async V All-Gather
Q projection
  -> wait for K/V
attention(local Q, global K/V)
```

K communication overlaps V and Q projection; V communication overlaps Q
projection. If these projection windows are longer than the corresponding
collective latency, communication is hidden from the critical path. This path
does not require a fused attention/projection operator.

The overlap interface is model-independent, while projection adapters are
model-specific. The first adapter splits BAGEL's packed generation-expert
projection by addressing its K, V, and Q weight slices independently. Quantized
projection implementations that cannot expose those slices fall back to the
non-overlapped path.

### 4. DiT gap parallelism

Attention CP alone shards the transformer-block region, but many integrations
gather the sequence at the model output and split it again on the next
denoising step. We call the region between two sharded transformer invocations
the **DiT gap**:

```text
sharded DiT output
  -> output projection / unpatchify
  -> CFG and post-processing
  -> scheduler update
  -> patchify / input projection
  -> next sharded DiT invocation
```

DiT gap parallelism keeps this region sharded and removes the redundant
per-step gather/re-split. It is a companion to KV-Gather CP: CP keeps attention
output local, while gap parallelism carries that local shard through the rest
of the denoising iteration.

Gap adaptation is model-specific:

- reshape/permutation patchify and unpatchify, Linear, LayerNorm/RMSNorm,
  AdaLN, element-wise CFG, and element-wise schedulers can remain local;
- global statistics such as GroupNorm or global CFG renormalization require
  All-Reduce of their sufficient statistics;
- spatial convolutions whose receptive fields cross shard boundaries require
  halo P2P exchange;
- model-specific conditioning, masking, or scheduler APIs must define whether
  their tensors remain sharded or are selectively reconstructed.

For a 2D latent with convolution in the DiT gap, the shard boundary should
preserve spatial topology. A natural layout splits H into horizontal stripes:
rank `i` owns `H_i` complete rows and therefore `H_i * W` image tokens. The
`H_i` values may differ by one row, which balances real latent tokens while
ensuring every spatial segment is a multiple of W. Halo P2P then exchanges
boundary rows between neighboring ranks.

Special or non-spatial tokens are not part of this divisibility constraint.
They should be represented separately from the spatial segment and may be:

- replicated when they are small and read-only;
- owned by rank 0 or the last rank when only one output copy is required; or
- assigned through explicit layout metadata when their outputs participate in
  later model-specific operations.

Replicated special K/V must stay outside the sequence All-Gather and be appended
once as joint attention K/V; otherwise the collective would duplicate those
tokens `P` times. An owner-sharded special token can participate in the regular
variable-length K/V gather because it appears on exactly one rank.

This separation allows local total sequence lengths to differ without breaking
the row-aligned layout required by spatial convolution.

BAGEL is the simplest first example. Its denoising state is already represented
as latent patch tokens, and its gap contains token-local `vae2llm`, `llm2vae`,
position/time embedding, CFG combination, and Euler update. It has no
GroupNorm, spatial convolution, patchify, or unpatchify inside the denoising
loop. Only global CFG renormalization needs an All-Reduce of local squared
norms. The implementation therefore splits the initial latent-token sequence
once and removes the full-velocity gather from every denoising step.

VAE encode/decode parallelism is a separate concern and is outside this RFC.
This proposal covers persistent sharding over the DiT denoising process.

### 5. Configuration

```text
--cp P
--context-parallel-degree P  # alias
```

The deploy configuration field is `context_parallel_degree`. Configuration
precedence is CLI, deploy YAML, `VLLM_OMNI_CP_DEGREE`, then 1.

### 6. Constraints

- The generic design supports variable local sequence lengths through cached
  valid-length metadata and padded K/V collectives.
- The first BAGEL implementation still uses equal latent shards and
  `all_gather_into_tensor`; variable-length support is follow-up implementation
  work rather than a design limitation.
- Attention masks and joint attention masks are not supported by the first
  BAGEL implementation.
- `context_parallel_degree > 1` is currently mutually exclusive with Ulysses
  and Ring degrees greater than 1.
- Full per-layer K/V must fit on every rank.
- GQA/MQA provides the strongest communication benefit; the formula above
  defines the crossover with Ulysses.
- The BAGEL CP path uses SDPA for denoising-trajectory alignment. FlashAttention
  is mathematically valid for the non-square local-Q/global-KV shape, but small
  per-layer differences can compound over transformer layers and denoising
  steps.
- Quantized QKV projections require support for independent K, V, and Q slices
  to use the asynchronous overlap fast path.
- DiT gap parallelism requires a model adapter for non-token-local operations.
- The BAGEL adapter currently falls back to the existing path for trajectory
  collection, custom SDE schedulers, and Lance frame conditioning.
- VAE encode/decode parallelism is not part of this design.

### 7. Validation

The first BAGEL implementation includes unit coverage for:

- gathering K/V while keeping Q local;
- configuration precedence and Ulysses/Ring mutual exclusion;
- independent generation-expert Q/K/V projection slices;
- globally correct CFG renormalization under CP;
- one final latent gather rather than one gather per denoising step.

The fixed BAGEL online CFG workload used 1024x1024 output, 50 steps, seed 52,
and CP/UP degree 4. On four NVIDIA H20 GPUs, measured wall time was 21.51 s for
Ulysses and 20.35 s for KV-Gather CP plus DiT gap parallelism. A trace confirmed
per-layer K/V All-Gather and removal of per-step velocity All-Gather.

Future model integrations should separately validate:

- local-Q/global-KV attention equivalence;
- variable-length K/V reconstruction and padding removal;
- projection/communication overlap;
- numerical equivalence of every non-token-local gap operation;
- topology-preserving spatial shards and halo exchange where convolution is
  present;
- absence of per-step gather/re-split in the denoising loop.

## Feedback Period

One week.

## Related Work

- BAGEL Ulysses/Ring sequence parallelism: #1903
- Continuous diffusion acceleration RFC: #1217

## Implementation

First model integration (BAGEL): #5315

**Comment by nussejzz** [comment]
@david6666666 @congw729 PTAL！
I am still working on it, but you can assign it to me first
